### PR TITLE
support magic link

### PR DESCRIPTION
**Pull request by helenren**

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java [hunk 1]
```diff
@@ -152,13 +157,34 @@ public static URI getAuthorizationUrl(URI loginServer, String clientId,
...
+        }
+        final StringBuilder sb = new StringBuilder(instanceURL);
+        sb.append(FRONTDOOR);
+        sb.append(SID).append(EQUAL).append(accessToken);
+        sb.append(AND).append(RETURL).append(EQUAL).append(Uri.encode(getAuthorizationUrlStr(loginServer,clientId,callbackUrl, scopes, clientSecret, displayType)));
+        return URI.create(sb.toString());
+    }
+
+
+    private static String getAuthorizationUrlStr(URI loginServer, String clientId,
```
Comment (wmathurin) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java]: Why do we need two methods? Couldn't we just have the one returning a URI, and if a string is wanted instead, the caller could call toString()?

File: external/shared [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 2f0a4b63b4bb7495c73f25af32647ad2d23a2282
+Subproject commit b08ba40ceded73b2b02fd4b1cf6df9308b352220
```
Comment (wmathurin) [inline external/shared]: Android's unstable points to 9ff3d on shared (which is more recent that b08ba)

Comment (helenren) [inline external/shared]: yes, need to update it and point to the latest one

File: build.gradle [hunk 3]
```diff
@@ -19,4 +19,4 @@ allprojects {
 
     jcenter()
   }
-}
\ No newline at end of file
+}
```
Comment (bhariharan) [inline build.gradle]: No changes in this file? Could you revert?

File: external/shared (hunk 2 above)
Comment (bhariharan) [inline external/shared]: Yeah, this needs to be rebased against ```unstable```.

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,6 +1,6 @@
-#Mon Apr 04 13:28:13 PDT 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-2.12-bin.zip
+
```
Comment (bhariharan) [inline gradle/wrapper/gradle-wrapper.properties]: Could you revert changes to this file?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java [hunk 5]
```diff
@@ -743,7 +745,14 @@ public LoginOptions(String loginUrl, String passcodeHash, String oauthCallbackUr
...
         }
 
+        public LoginOptions(String loginUrl, String passcodeHash, String oauthCallbackUrl, String oauthClientId, String[] oauthScopes, String clientSecret, String jwt) {
+            this(loginUrl, passcodeHash, oauthCallbackUrl, oauthClientId, oauthScopes, clientSecret);
+            this.jwt = jwt;
+            bundle.putString(JWT, jwt);
+        }
+
         public Bundle asBundle() {
+            bundle.putString(JWT, jwt); //jwt can be updated during flow
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java]: I don't think we should muck around with this in this part.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java [hunk 6]
```diff
@@ -248,17 +248,27 @@ protected void showError(Exception exception) {
...
+            // Filling in loginUrl.
+            loginOptions.loginUrl = getLoginUrl();
+        }
+        if (loginOptions.jwt != null) {
+            new SwapJWTForAccessTokenTask().execute(loginOptions);
+        }
+        this.doLoadPage(false);
+    }
+
+    private void doLoadPage(Boolean jwtFlow) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java]: @helenren Use lowercase ```boolean```.

**Comment by helenren** [comment]
@bhariharan @wmathurin updated the codes based on your comments

**Comment by wmathurin** [comment]
LGTM

File: build.gradle [hunk 7]
```diff
@@ -10,7 +10,7 @@ buildscript {
 
 allprojects {
   group = 'com.salesforce.androidsdk'
-  
+
```
Comment (bhariharan) [inline build.gradle]: Could you revert this file? Nothing has changed here.

File: gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.12-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.12-bin.zip
```
Comment (bhariharan) [inline gradle/wrapper/gradle-wrapper.properties]: Could you revert this file? Nothing has changed here.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java [hunk 9]
```diff
@@ -263,13 +263,32 @@ public void switchToNewUser() {
...
+	 * Kicks off the login flow to switch to a new user with jwt. Once the login
+	 * flow is complete, the context will automatically become the
+	 * new user's context and a call to peekRestClient() or getRestClient()
+	 * in ClientManager will return a RestClient instance for the new user.
+	 *
+	 * @param jwt JWT.
+	 * @param url Instance/My domain URL.
+	 */
+	public void switchToNewUser(String jwt, String url) {
+		final Bundle reply = new Bundle();
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java]: @helenren A lot of the code here is the same as the previous method, except for one line that builds ```options```. Let's create one common ```private``` method that does all the common stuff that takes in a ```Bundle``` - ```options```.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java [hunk 10]
```diff
@@ -275,14 +275,27 @@ public SalesforceR getSalesforceR() {
...
-		if (loginOptions == null) {
-			final BootConfig config = BootConfig.getBootConfig(context);
-			loginOptions = new LoginOptions(null, getPasscodeHash(), config.getOauthRedirectURI(),
-	        		config.getRemoteAccessConsumerKey(), config.getOauthScopes());
-		}
-		return loginOptions;
+		return getLoginOptionsInternal(null, null);
 	}
 
+    public LoginOptions getLoginOptions(String jwt, String url) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java]: Remove this.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java [hunk 11]
```diff
@@ -275,14 +275,27 @@ public SalesforceR getSalesforceR() {
...
+        return getLoginOptionsInternal(jwt, url);
+    }
+
+    public LoginOptions getLoginOptionsInternal(String jwt, String url) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java]: Rename to ```getLoginOptions```.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java [hunk 12]
```diff
@@ -260,12 +260,34 @@ public void switchToUser(UserAccount user) {
...
+	/**
+	 * Kicks off the login flow to switch to a new user with jwt. Once the login
+	 * flow is complete, the context will automatically become the
+	 * new user's context and a call to peekRestClient() or getRestClient()
+	 * in ClientManager will return a RestClient instance for the new user.
+	 *
+	 * @param jwt JWT.
+	 * @param url Instance/My domain URL.
+	 */
+	public void switchToNewUser(String jwt, String url) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java]: Based on what we discussed offline, let's pass in ```Bundle```.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java [hunk 13]
```diff
@@ -260,12 +260,34 @@ public void switchToUser(UserAccount user) {
...
+		switchToNewUserInternal(jwt, url);
+	}
+
+	private void switchToNewUserInternal (String jwt, String url) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java]: Rename to ```switchToNewUserWithOptions```.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java [hunk 14]
```diff
@@ -256,6 +277,28 @@ public static TokenEndpointResponse swapAuthCodeForTokens(HttpAccess httpAccesso
...
+     *
+     */
+    public static TokenEndpointResponse swapJWTForTokens(HttpAccess httpAccessor, URI loginServerUrl,
+                                                              String jwt) throws IOException, URISyntaxException, OAuthFailedException {
+        // call the token endpoint, and swap jwt for an access tokens.
+        FormBody.Builder formBodyBuilder = new FormBody.Builder()
+                .add(GRANT_TYPE, JWT_BEARER)
+                .add(ASSERTION, jwt);
+        TokenEndpointResponse tr = makeTokenEndpointRequest(httpAccessor, loginServerUrl, formBodyBuilder);
+        return tr;
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java]: ```return``` directly.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java [hunk 15]
```diff
@@ -255,10 +255,19 @@ protected void showError(Exception exception) {
      */
     public void loadLoginPage() {
 
-        // Filling in loginUrl.
-        loginOptions.loginUrl = getLoginUrl();
+        if (loginOptions.jwt == null || loginOptions.loginUrl == null) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java]: This should be:

```if (TextUtils.isEmpty(loginOptions.jwt)) {
loginOptions.loginUrl = getLoginUrl();
doLoadPage(false);
} else {
new SwapJWTForAccessTokenTask().execute(loginOptions);
}```

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/accounts/SmartSyncUserAccountManager.java [hunk 16]
```diff
@@ -63,8 +63,21 @@ public void switchToUser(UserAccount user) {
 
 	@Override
 	public void switchToNewUser() {
-		super.switchToNewUser();
-    	final UserAccount userAccount = SmartSyncUserAccountManager.getInstance().getCurrentUser();
+		switchToNewUserInternal(null, null);
+	}
+
+	@Override
+	public void switchToNewUser(String jwt, String url) {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/accounts/SmartSyncUserAccountManager.java]: Remove this.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/accounts/SmartSyncUserAccountManager.java [hunk 17]
```diff
@@ -63,8 +63,21 @@ public void switchToUser(UserAccount user) {
...
+		switchToNewUserInternal(jwt, url);
+	}
+
+	private void switchToNewUserInternal(String jwt, String url) {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/accounts/SmartSyncUserAccountManager.java]: Rename to ```switchToNewUser```.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java (hunk 12 above)
Comment (helenren) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java]: addressed

**Comment by helenren** [comment]
Also tested locally.

**Comment by bhariharan** [comment]
👍

**Comment by kchitalia** [comment]
Did we add any tests around this flow?

**Comment by bhariharan** [comment]
@kchitalia We don't have tests around the ```OAuth``` flow right now, since it's in the ```WebView``` and hard to automate. We do have a story to spike on options in the future though.

**Comment by kchitalia** [comment]
:+1:

**Pull request merged**